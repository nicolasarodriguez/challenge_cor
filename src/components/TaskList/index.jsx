import React from 'react'
import { array, func, string } from 'prop-types'

import Card from 'react-bootstrap/Card'

import TaskData from '../TaskData'
import EmptyState from '../EmptyState'

const namespace = 'ui-tasklist'

const TaskList = ({ list, setList, priorityFilter, statusFilter }) => {
  const onChangeHandler = (event, id, property) => {
    const { value } = event.target
    const objIndex = list.findIndex((obj => obj.id === id));
    list[objIndex][property] = value
    setList([...list])
  }

  const deleteTask = (id) => {
    const objIndex = list.findIndex((obj => obj.id === id));
    list.splice(objIndex, 1);
    setList([...list])
    
  }

  const todoItems = list.filter(item => (
    !((statusFilter !== 'all' && item.status !== statusFilter) ||
    (priorityFilter !== 'all' && item.priority !== priorityFilter))
  )).map(filteredItem => {
    return (
      <TaskData
        key={filteredItem.id}
        data={filteredItem}
        onChangeHandler={onChangeHandler}
        deleteTask={deleteTask}
      />
    )
  })

  const renderResult = () => {
    if (list.length > 0) {
      if (todoItems.length > 0) {
        return todoItems
      } else {
        return (<EmptyState text="No se encontro ningun resultado" image="no-results.jpg" />)
      }
    } else {
      return (<EmptyState text="No hay tareas cargadas" image="no-tasks.jpg" />)
    }
  }

  return (
    <Card className={`${namespace}`}>
      <Card.Body>
        {renderResult()}
      </Card.Body>
    </Card>
  )
}

TaskList.propTypes = {
  list: array.isRequired,
  setList: func.isRequired,
  priorityFilter: string.isRequired,
  statusFilter: string.isRequired
}

export default TaskList
