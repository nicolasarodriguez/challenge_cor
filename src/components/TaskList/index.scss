@import '../../styles/variales.scss';

.tasklist {
    background-color: rgb(255, 255, 255);    
    border-radius: 4px;
    box-shadow: rgba(0, 0, 0, 0.25) 0px 1px 2px 0px;
    margin: 0 auto 20px;
    padding: 20px;
    max-width: $max-width;
}