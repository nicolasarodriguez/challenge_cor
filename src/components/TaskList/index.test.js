/* eslint-disable testing-library/no-node-access */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import TaskList from './index';

const defaultProps = {
    list: [{
        description: "Descripcion",
        id: "1648826951727",
        priority: "HIGH",
        status: "NEW",
        title: "Titulo"
    }],
    setList: jest.fn(),
    priorityFilter: 'all',
    statusFilter: 'NEW',
}

const setup = (propOverrides = {}) => {
    const props = {
      ...defaultProps,
      ...propOverrides,
    };
  
    const view = render(<TaskList {...props} />);
  
    return {
      props,
      view
    };
};

describe('<TaskList />', () => {
    it('should match snapshot', () => {
        const { view } = setup();
        const title = screen.getByText('Titulo');
        expect(title).toBeInTheDocument();
        expect(view).toMatchSnapshot();
    });

    it('should match snapshot with no tasks', () => {
        const { view } = setup({ list: [] });
        const title = screen.getByText('No hay tareas cargadas');
        expect(title).toBeInTheDocument();
        expect(view).toMatchSnapshot();
    });

    it('should match snapshot with no results', () => {
        const { view } = setup({ priorityFilter: 'LOW', });
        const title = screen.getByText('No se encontro ningun resultado');
        expect(title).toBeInTheDocument();
        expect(view).toMatchSnapshot();
    });
});
