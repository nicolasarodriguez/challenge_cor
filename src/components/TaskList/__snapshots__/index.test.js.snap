// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskList /> should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui-tasklist card"
      >
        <div
          class="card-body"
        >
          <div
            class="ui-taskdata ui-taskdata--high card"
          >
            <div
              class="card-body"
            >
              <div
                class="ui-taskdata__info"
              >
                <div>
                  <p>
                    <span>
                      Prioridad:
                    </span>
                    <select>
                      <option
                        value="HIGH"
                      >
                        Alta
                      </option>
                      <option
                        value="MEDIUM"
                      >
                        Media
                      </option>
                      <option
                        value="LOW"
                      >
                        Baja
                      </option>
                    </select>
                  </p>
                </div>
                <div>
                  <p>
                    <span>
                      Estado:
                    </span>
                    <select>
                      <option
                        value="NEW"
                      >
                        Nueva
                      </option>
                      <option
                        value="IN_PROGRESS"
                      >
                        En proceso
                      </option>
                      <option
                        value="DONE"
                      >
                        Finalizada
                      </option>
                    </select>
                  </p>
                </div>
                <div
                  class="ui-taskdata__button-container"
                >
                  <button
                    class="btn btn-danger"
                    type="button"
                  >
                    Eliminar
                  </button>
                </div>
              </div>
              <span
                class="ui-taskdata__id"
              >
                #
                1648826951727
              </span>
              <h2
                class="ui-taskdata__title"
              >
                Titulo
              </h2>
              <p>
                Descripcion
                <br />
                <span>
                  Descripcion
                </span>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui-tasklist card"
    >
      <div
        class="card-body"
      >
        <div
          class="ui-taskdata ui-taskdata--high card"
        >
          <div
            class="card-body"
          >
            <div
              class="ui-taskdata__info"
            >
              <div>
                <p>
                  <span>
                    Prioridad:
                  </span>
                  <select>
                    <option
                      value="HIGH"
                    >
                      Alta
                    </option>
                    <option
                      value="MEDIUM"
                    >
                      Media
                    </option>
                    <option
                      value="LOW"
                    >
                      Baja
                    </option>
                  </select>
                </p>
              </div>
              <div>
                <p>
                  <span>
                    Estado:
                  </span>
                  <select>
                    <option
                      value="NEW"
                    >
                      Nueva
                    </option>
                    <option
                      value="IN_PROGRESS"
                    >
                      En proceso
                    </option>
                    <option
                      value="DONE"
                    >
                      Finalizada
                    </option>
                  </select>
                </p>
              </div>
              <div
                class="ui-taskdata__button-container"
              >
                <button
                  class="btn btn-danger"
                  type="button"
                >
                  Eliminar
                </button>
              </div>
            </div>
            <span
              class="ui-taskdata__id"
            >
              #
              1648826951727
            </span>
            <h2
              class="ui-taskdata__title"
            >
              Titulo
            </h2>
            <p>
              Descripcion
              <br />
              <span>
                Descripcion
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<TaskList /> should match snapshot with no results 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui-tasklist card"
      >
        <div
          class="card-body"
        >
          <div
            class="ui-emptystate"
          >
            <img
              alt="empty state"
              class="ui-emptystate__image"
              src="/assets/no-results.jpg"
            />
            <p
              class="ui-emptystate__text"
            >
              No se encontro ningun resultado
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui-tasklist card"
    >
      <div
        class="card-body"
      >
        <div
          class="ui-emptystate"
        >
          <img
            alt="empty state"
            class="ui-emptystate__image"
            src="/assets/no-results.jpg"
          />
          <p
            class="ui-emptystate__text"
          >
            No se encontro ningun resultado
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<TaskList /> should match snapshot with no tasks 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui-tasklist card"
      >
        <div
          class="card-body"
        >
          <div
            class="ui-emptystate"
          >
            <img
              alt="empty state"
              class="ui-emptystate__image"
              src="/assets/no-tasks.jpg"
            />
            <p
              class="ui-emptystate__text"
            >
              No hay tareas cargadas
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui-tasklist card"
    >
      <div
        class="card-body"
      >
        <div
          class="ui-emptystate"
        >
          <img
            alt="empty state"
            class="ui-emptystate__image"
            src="/assets/no-tasks.jpg"
          />
          <p
            class="ui-emptystate__text"
          >
            No hay tareas cargadas
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
